app/models/specific_tube_creation.rb

```diff
@@ -64,13 +64,41 @@ def connect_parent_and_children
   end
   private :connect_parent_and_children
 
+  def determine_foreign_barcode_format(foreign_barcode)
+    Barcode.matching_barcode_format(foreign_barcode)
+  end
+
+  def check_foreign_barcode_unique(foreign_barcode_format, foreign_barcode)
+    return unless Barcode.exists_for_format?(foreign_barcode_format, foreign_barcode)
+
+    raise "Foreign Barcode: #{foreign_barcode} is already in use!"
```

Comment:
This should probably be validation, rather than an exception. Also, it may be better to just use .valid? on barcode, and extract the errors from that, rather than duplicating logic all over the place.